itou/utils/address/format.py

```diff
@@ -0,0 +1,112 @@
+from unidecode import unidecode
+
+from itou.asp.models import LaneExtension, LaneType, find_lane_type_aliases
+from itou.users.models import User
+from itou.utils.apis.geocoding import get_geocoding_data
+
+
+def format_address(obj, update_coords=False, strict=True):
+    """
+    Formats the address contained in obj into a valid address "structure" for ASP ER exports.
+
+    Heavily relies on geo.api.gouv.fr API to do parts of the job for us:
+    - extracting lane number and extension
+    - giving a geocoding score / threshold in order to improve an existing DB address
+    - validation of a real-world address
+
+    Employee records ("Fiches salarié") contains 2 addresses of this kind.
+
+    See validation of ASP address for expected/valid fields.
+
+    Output fields:
+    - number (opt.): number in the lane
+    - std_extension (opt.): One of the ASP ref lane extension (see LaneExtension)
+    - non_std_extension (opt.): if another extension is detected
+    - lane: name of the lane
+    - lane_type: One of the ASP ref lane type (see LaneType)
+    - city: name of city
+    - post_code: postal code
+    - insee_code: INSEE code of the city (Itou)
+
+    INSEE code can be checked against ASP ref for further validation.
+
+    Returns a (result,error) tuple:
+    - OK => (result_dict, None),
+    - KO => (None, error_message)
+    """
+    if strict and not isinstance(obj, User):
+        return None, "Only valid for User objects"
+
+    # Do we have enough data to make an extraction?
+    if not obj.post_code or not obj.address_line_1:
+        return None, "Incomplete address data"
+
+    # first we use geo API to get a 'lane' and a number
+    address = get_geocoding_data(obj.address_line_1, post_code=obj.post_code)
+
+    if not address:
+        return None, "Geocoding error, unable to get result"
+
+    result = {}
+
+    # Street extension processing (bis, ter ...)
+    # Extension is part of the resulting streetnumber geo API field
+    number_plus_ext = address.get("number")
+    if number_plus_ext:
+        number, *extension = number_plus_ext.split()
```

Comment:
TIL [Asterisks in tuple unpacking](https://treyhunner.com/2018/10/asterisks-in-python-what-they-are-and-how-to-use-them/#Asterisks_in_tuple_unpacking) 👍